In [1]:
import pandas as pd
import numpy as np

In [2]:
df_country= pd.read_excel("//Users/pkc/PycharmProjects/pythonProject/Automotive/data/Car-Models-List-by-Teoalida-SAMPLE.xlsx",sheet_name='car_make_list')

In [3]:
df_country.head(20)

Unnamed: 0                                         Unnamed: 1   
0          NaN                                               Make  \
1          NaN                                                NaN   
2          NaN                                                222   
3          NaN                                                NaN   
4          NaN                             CAR MANUFACTURERS LIST   
5          NaN  Original research by Teoalida © cardatabase.te...   
6          NaN                                                NaN   
7          NaN  This is NOT a list of all manufacturers ever e...   
8          NaN  If you saw somewhere a list with 1000 manufact...   
9          NaN  NOTE: Sold in Western Europe / North America i...   
10         NaN                                                NaN   
11         NaN  Car makes with list of models included (and nu...   
12         NaN                                                NaN   
13         NaN                                            AC Cars   
14         NaN                                              Acura   
15         NaN                       Aixam-Mega (including Arola)   
16         NaN                                         Alfa Romeo   
17         NaN                                             Alpine   
18         NaN                                              Alvis   
19         NaN                                         AMC, Eagle   

        Unnamed: 2 Unnamed: 3               Unnamed: 4   
0          Country       Size  Models included (World)  \
1              NaN        NaN                      NaN   
2              222        222                     6264   
3              NaN        NaN                      NaN   
4              NaN        NaN                      NaN   
5              NaN        NaN                      NaN   
6              NaN        NaN                      NaN   
7              NaN        NaN                      NaN   
8              NaN        NaN                      NaN   
9              NaN        NaN                      NaN   
10             NaN        NaN                      NaN   
11             NaN        NaN                      NaN   
12             NaN        NaN                      NaN   
13  United Kingdom      Small                       11   
14           Japan      Large                       37   
15          France      Small                       32   
16           Italy      Large                       35   
17          France     Medium                        7   
18  United Kingdom     Medium                        9   
19   United States      Large                       33   

                  Unnamed: 5         Unnamed: 6       Unnamed: 7   
0   Models included (Europe)  Timeline included  Company Founded  \
1                        NaN                NaN              NaN   
2                       3320                104              198   
3                        NaN                NaN              NaN   
4                        NaN                NaN              NaN   
5                        NaN                NaN              NaN   
6                        NaN                NaN              NaN   
7                        NaN                NaN              NaN   
8                        NaN                NaN              NaN   
9                        NaN                NaN              NaN   
10                       NaN                NaN              NaN   
11                       NaN                NaN              NaN   
12                       NaN                NaN              NaN   
13                        11                YES             1901   
14                       NaN                YES             1986   
15                        32                NaN       1975, 1983   
16                        35                YES             1910   
17                         7                YES             1955   
18                         9               

imported looks terrible, need to make first row the column list, and remove like the first ten rows which are text, not data.

In [4]:
df_country.columns= df_country.iloc[0]

In [5]:
df_country.head(15)

0  NaN                                               Make         Country   
0  NaN                                               Make         Country  \
1  NaN                                                NaN             NaN   
2  NaN                                                222             222   
3  NaN                                                NaN             NaN   
4  NaN                             CAR MANUFACTURERS LIST             NaN   
5  NaN  Original research by Teoalida © cardatabase.te...             NaN   
6  NaN                                                NaN             NaN   
7  NaN  This is NOT a list of all manufacturers ever e...             NaN   
8  NaN  If you saw somewhere a list with 1000 manufact...             NaN   
9  NaN  NOTE: Sold in Western Europe / North America i...             NaN   
10 NaN                                                NaN             NaN   
11 NaN  Car makes with list of models included (and nu...             NaN   
12 NaN                                                NaN             NaN   
13 NaN                                            AC Cars  United Kingdom   
14 NaN                                              Acura           Japan   

0    Size  Models included (World)  Models included (Europe)   
0    Size  Models included (World)  Models included (Europe)  \
1     NaN                      NaN                       NaN   
2     222                     6264                      3320   
3     NaN                      NaN                       NaN   
4     NaN                      NaN                       NaN   
5     NaN                      NaN                       NaN   
6     NaN                      NaN                       NaN   
7     NaN                      NaN                       NaN   
8     NaN                      NaN                       NaN   
9     NaN                      NaN                       NaN   
10    NaN                      NaN                       NaN   
11    NaN                      NaN                       NaN   
12    NaN                      NaN                       NaN   
13  Small                       11                        11   
14  Large                       37                       NaN   

0   Timeline included  Company Founded  First car produced  Last car produced   
0   Timeline included  Company Founded  First car produced  Last car produced  \
1                 NaN              NaN                 NaN                NaN   
2                 104              198                 178                222   
3                 NaN              NaN                 NaN                NaN   
4                 NaN              NaN                 NaN                NaN   
5                 NaN              NaN                 NaN                NaN   
6                 NaN              NaN                 NaN                NaN   
7                 NaN              NaN                 NaN                NaN   
8                 NaN              NaN                 NaN                NaN   
9                 NaN              NaN                 NaN                NaN   
10                NaN              NaN                 NaN                NaN   
11                NaN              NaN                 NaN                NaN   
12                NaN              NaN                 NaN                NaN   
13                YES             1901                1903             Active   
14                YES             1986                1986             Active   

0   ...    NaN    NaN  Sold in North America    NaN    NaN    NaN    NaN   
0   ...    NaN    NaN  Sold in North America    NaN    NaN    NaN    NaN  \
1   ...  2010s  2020s                  1960s  1970s  1980s  1990s  2000s   
2   ...     55     50                     54     57     56     51     54   
3   ...    NaN    NaN                    NaN    NaN    NaN    NaN    NaN   
4   ...    NaN    NaN                    NaN    NaN    NaN    NaN

In [6]:
df_country[13:].head()

0  NaN                          Make         Country    Size   
13 NaN                       AC Cars  United Kingdom   Small  \
14 NaN                         Acura           Japan   Large   
15 NaN  Aixam-Mega (including Arola)          France   Small   
16 NaN                    Alfa Romeo           Italy   Large   
17 NaN                        Alpine          France  Medium   

0  Models included (World) Models included (Europe) Timeline included   
13                      11                       11               YES  \
14                      37                      NaN               YES   
15                      32                       32               NaN   
16                      35                       35               YES   
17                       7                        7               YES   

0  Company Founded First car produced Last car produced  ...  NaN  NaN   
13            1901               1903            Active  ...  NaN  NaN  \
14            1986               1986            Active  ...  NaN  NaN   
15      1975, 1983                NaN            Active  ...  NaN  NaN   
16            1910               1910            Active  ...    Y    Y   
17            1955               1955            Active  ...  NaN  NaN   

0  Sold in North America  NaN  NaN  NaN  NaN  NaN  NaN   
13                     Y    Y    Y    Y  NaN  NaN  NaN  \
14                   NaN  NaN  NaN    Y    Y    Y    Y   
15                   NaN  NaN  NaN  NaN  NaN  NaN  NaN   
16                     Y    Y    Y    Y  NaN    Y    Y   
17                   NaN  NaN  NaN  NaN  NaN  NaN  NaN   

0                                         Description  
13                                                NaN  
14                              Honda division for US  
15                                                NaN  
16                                                NaN  
17  Bought by Renault in 1978, defunct 1995, relau...  

[5 rows x 26 columns]

In [7]:
df_country= df_country[13:].copy()

In [8]:
df_country.head()

0  NaN                          Make         Country    Size   
13 NaN                       AC Cars  United Kingdom   Small  \
14 NaN                         Acura           Japan   Large   
15 NaN  Aixam-Mega (including Arola)          France   Small   
16 NaN                    Alfa Romeo           Italy   Large   
17 NaN                        Alpine          France  Medium   

0  Models included (World) Models included (Europe) Timeline included   
13                      11                       11               YES  \
14                      37                      NaN               YES   
15                      32                       32               NaN   
16                      35                       35               YES   
17                       7                        7               YES   

0  Company Founded First car produced Last car produced  ...  NaN  NaN   
13            1901               1903            Active  ...  NaN  NaN  \
14            1986               1986            Active  ...  NaN  NaN   
15      1975, 1983                NaN            Active  ...  NaN  NaN   
16            1910               1910            Active  ...    Y    Y   
17            1955               1955            Active  ...  NaN  NaN   

0  Sold in North America  NaN  NaN  NaN  NaN  NaN  NaN   
13                     Y    Y    Y    Y  NaN  NaN  NaN  \
14                   NaN  NaN  NaN    Y    Y    Y    Y   
15                   NaN  NaN  NaN  NaN  NaN  NaN  NaN   
16                     Y    Y    Y    Y  NaN    Y    Y   
17                   NaN  NaN  NaN  NaN  NaN  NaN  NaN   

0                                         Description  
13                                                NaN  
14                              Honda division for US  
15                                                NaN  
16                                                NaN  
17  Bought by Renault in 1978, defunct 1995, relau...  

[5 rows x 26 columns]

In [9]:
df_country.columns

Index([                          nan,                        'Make',
                           'Country',                        'Size',
           'Models included (World)',    'Models included (Europe)',
                 'Timeline included',             'Company Founded',
                'First car produced',           'Last car produced',
       'Model Years (North America)',      'Sold in Western Europe',
                                 nan,                           nan,
                                 nan,                           nan,
                                 nan,                           nan,
             'Sold in North America',                           nan,
                                 nan,                           nan,
                                 nan,                           nan,
                                 nan,                 'Description'],
      dtype='object', name=0)

In [10]:
df_country.reset_index()

0    index NaN                                  Make         Country    Size   
0       13 NaN                               AC Cars  United Kingdom   Small  \
1       14 NaN                                 Acura           Japan   Large   
2       15 NaN          Aixam-Mega (including Arola)          France   Small   
3       16 NaN                            Alfa Romeo           Italy   Large   
4       17 NaN                                Alpine          France  Medium   
..     ...  ..                                   ...             ...     ...   
234    247 NaN              Austin (included in BMC)  United Kingdom   Large   
235    248 NaN               Riley (included in BMC)  United Kingdom   Large   
236    249 NaN            Wolseley (included in BMC)  United Kingdom   Large   
237    250 NaN            Standard (included in BMC)  United Kingdom   Large   
238    251 NaN  Talbot (included in Chrysler Europe)  United Kingdom   Large   

0   Models included (World) Models included (Europe) Timeline included   
0                        11                       11               YES  \
1                        37                      NaN               YES   
2                        32                       32               NaN   
3                        35                       35               YES   
4                         7                        7               YES   
..                      ...                      ...               ...   
234                       -                        -               NaN   
235                       -                        -               NaN   
236                       -                        -               NaN   
237                       -                        -               NaN   
238                       -                        -               NaN   

0   Company Founded First car produced  ...  NaN  NaN Sold in North America   
0              1901               1903  ...  NaN  NaN                     Y  \
1              1986               1986  ...  NaN  NaN                   NaN   
2        1975, 1983                NaN  ...  NaN  NaN                   NaN   
3              1910               1910  ...    Y    Y                     Y   
4              1955               1955  ...  NaN  NaN                   NaN   
..              ...                ...  ...  ...  ...                   ...   
234            1905               1906  ...  NaN  NaN                     Y   
235            1896                NaN  ...  NaN  NaN                     Y   
236            1901                NaN  ...  NaN  NaN                   NaN   
237            1903                NaN  ...  NaN  NaN                   NaN   
238            1903                NaN  ...  NaN  NaN                   NaN   

0    NaN  NaN  NaN  NaN  NaN  NaN   
0      Y    Y    Y  NaN  NaN  NaN  \
1    NaN  NaN    Y    Y    Y    Y   
2    NaN  NaN  NaN  NaN  NaN  NaN   
3      Y    Y    Y  NaN    Y    Y   
4    NaN  NaN  NaN  NaN  NaN  NaN   
..   ...  ...  ...  ...  ...  ...   
234    Y  NaN  NaN  NaN  NaN  NaN   
235  NaN  NaN  NaN  NaN  NaN  NaN   
236  NaN  NaN  NaN  NaN  NaN  NaN   
237  NaN  NaN  NaN  NaN  NaN  NaN   
238  NaN  NaN  NaN  NaN  NaN  NaN   

0                                          Description  
0                                                  NaN  
1                                Honda division for US  
2                                                  NaN  
3                                                  NaN  
4    Bought by Renault in 1978, defunct 1995, relau...  
..                                                 ...  
234   Merged in British Leyland, brand used until 1994  
235   Merged in British Leyland, brand used until 1969  
236   Merged in British Leyland, brand used until 1975  
237   Merged in British Leyland, brand used until 1963  
238  Revived by PSA in 1979, passager cars disconti...  

[239 rows x 27 columns]

In [11]:
df_country= df_country.reset_index()

In [12]:
df_make_origin= df_country[['Make','Country']].copy()

In [13]:
df_make_origin.head()

0                          Make         Country
0                       AC Cars  United Kingdom
1                         Acura           Japan
2  Aixam-Mega (including Arola)          France
3                    Alfa Romeo           Italy
4                        Alpine          France

In [14]:
df_make_origin.rename(columns={'Make':'make','Country':'country'},inplace=True)

In [15]:
df_make_origin.head()

0                          make         country
0                       AC Cars  United Kingdom
1                         Acura           Japan
2  Aixam-Mega (including Arola)          France
3                    Alfa Romeo           Italy
4                        Alpine          France

In [16]:
df_make_origin.make.value_counts()

make
Buick                                   2
Audi                                    2
AC Cars                                 1
Plymouth                                1
Opel                                    1
                                       ..
Hindustan                               1
Holden                                  1
Honda                                   1
Honda India                             1
Talbot (included in Chrysler Europe)    1
Name: count, Length: 235, dtype: int64

In [17]:
df_make_origin[(df_make_origin.make=='Buick')|(df_make_origin.make=='Audi')]

0     make        country
13    Audi        Germany
28   Buick  United States
29   Buick          China
225   Audi        Germany

In [18]:
df_make_origin= df_make_origin.drop(29)

In [19]:
df_make_origin= df_make_origin.drop(225)

In [20]:
df_autos= pd.read_pickle('/Users/pkc/PycharmProjects/pythonProject/Automotive/pickles/df_autos.pickle')

In [21]:
df_autos.shape

(46405, 9)

In [22]:
df_autos.head(10)

mileage        make    model               fuel       gear offerType   
0   235000         BMW      316             Diesel     Manual      Used  \
1    92800  Volkswagen     Golf           Gasoline     Manual      Used   
2   149300        SEAT     Exeo           Gasoline     Manual      Used   
3    96200     Renault   Megane           Gasoline     Manual      Used   
4   156000     Peugeot      308           Gasoline     Manual      Used   
5   147000      Toyota    Auris  Electric/Gasoline  Automatic      Used   
6    91894     Renault   Scenic             Diesel     Manual      Used   
7   127500        Opel   Zafira           Gasoline     Manual      Used   
8   115000       Mazda        3           Gasoline  Automatic      Used   
9      104        Ford  Transit             Diesel     Manual      Used   

   price     hp  year  
0   6800  116.0  2011  
1   6877  122.0  2011  
2   6900  160.0  2011  
3   6950  110.0  2011  
4   6950  156.0  2011  
5   6950   99.0  2011  
6   6970  131.0  2011  
7   6972  116.0  2011  
8   6980  150.0  2011  
9   6990   86.0  2011

In [23]:
df_autos.head(10)

mileage        make    model               fuel       gear offerType   
0   235000         BMW      316             Diesel     Manual      Used  \
1    92800  Volkswagen     Golf           Gasoline     Manual      Used   
2   149300        SEAT     Exeo           Gasoline     Manual      Used   
3    96200     Renault   Megane           Gasoline     Manual      Used   
4   156000     Peugeot      308           Gasoline     Manual      Used   
5   147000      Toyota    Auris  Electric/Gasoline  Automatic      Used   
6    91894     Renault   Scenic             Diesel     Manual      Used   
7   127500        Opel   Zafira           Gasoline     Manual      Used   
8   115000       Mazda        3           Gasoline  Automatic      Used   
9      104        Ford  Transit             Diesel     Manual      Used   

   price     hp  year  
0   6800  116.0  2011  
1   6877  122.0  2011  
2   6900  160.0  2011  
3   6950  110.0  2011  
4   6950  156.0  2011  
5   6950   99.0  2011  
6   6970  131.0  2011  
7   6972  116.0  2011  
8   6980  150.0  2011  
9   6990   86.0  2011

In [24]:
df_joined_countries= pd.merge(df_autos,df_make_origin,left_on='make',right_on='make',how='left')

In [25]:
df_joined_countries.head(10)

mileage        make    model               fuel       gear offerType   
0   235000         BMW      316             Diesel     Manual      Used  \
1    92800  Volkswagen     Golf           Gasoline     Manual      Used   
2   149300        SEAT     Exeo           Gasoline     Manual      Used   
3    96200     Renault   Megane           Gasoline     Manual      Used   
4   156000     Peugeot      308           Gasoline     Manual      Used   
5   147000      Toyota    Auris  Electric/Gasoline  Automatic      Used   
6    91894     Renault   Scenic             Diesel     Manual      Used   
7   127500        Opel   Zafira           Gasoline     Manual      Used   
8   115000       Mazda        3           Gasoline  Automatic      Used   
9      104        Ford  Transit             Diesel     Manual      Used   

   price     hp  year  country  
0   6800  116.0  2011  Germany  
1   6877  122.0  2011  Germany  
2   6900  160.0  2011      NaN  
3   6950  110.0  2011   France  
4   6950  156.0  2011   France  
5   6950   99.0  2011    Japan  
6   6970  131.0  2011   France  
7   6972  116.0  2011  Germany  
8   6980  150.0  2011    Japan  
9   6990   86.0  2011      NaN

In [26]:
df_joined_countries.country.isnull().value_counts()

country
False    35732
True     10673
Name: count, dtype: int64

In [27]:
df_joined_countries[df_joined_countries.country.isnull()].make.value_counts()

make
Ford                4442
Mercedes-Benz       2354
SEAT                1924
Citroen              957
MINI                 469
Land                 164
Alfa                 132
Cupra                 65
Abarth                43
Aston                 30
Others                25
DS                    16
Microcar              12
Alpina                10
Trailer-Anhänger       4
Baic                   3
Caravans-Wohnm         3
Aixam                  3
Maybach                3
Corvette               3
RAM                    2
DFSK                   2
Estrima                2
9ff                    1
Zhidou                 1
Trucks-Lkw             1
Brilliance             1
FISKER                 1
Name: count, dtype: int64

In [28]:
df_make_origin.make[50:75]

51            DeSoto
52         De Tomaso
53             Dodge
54       Donkervoort
55             Edsel
56       e.GO Mobile
57    Eicher Polaris
58              Elva
59        Facel Vega
60           Ferrari
61              Fiat
62    Fiat do Brasil
63        Fiat India
64     Fisker, Karma
65      Ford America
66    Ford Australia
67    Ford do Brasil
68       Ford Europe
69        Ford India
70      Freightliner
71               FSO
72               GAZ
73             Geely
74               Geo
75           Gilbern
Name: make, dtype: object

In [29]:
df_autos['make']=df_autos['make'].replace(to_replace='Cupra',value='SEAT')

In [30]:
df_make_origin["make"]= df_make_origin['make'].replace(to_replace='Ford America',value="Ford")

In [31]:
df_make_origin[131:132]

0       make        country
132  Mercury  United States

In [32]:
df_make_origin['make']=df_make_origin['make'].replace(to_replace="Mercedes-Benz (including Maybach)",value='Mercedes-Benz')

In [33]:
df_make_origin['make']=df_make_origin['make'].replace(to_replace="Seat",value='SEAT')

In [34]:
df_make_origin['make']=df_make_origin['make'].replace(to_replace="Mini",value='MINI')

In [35]:
df_make_origin['make']=df_make_origin['make'].replace(to_replace="Citroën",value='Citroen')

In [36]:
df_make_origin['make']=df_make_origin['make'].replace(to_replace="Land Rover",value='Land')

In [37]:
df_make_origin['make']=df_make_origin['make'].replace(to_replace="Aston Martin",value='Aston')

In [38]:
df_make_origin['make']=df_make_origin['make'].replace(to_replace="Alfa Romeo",value='Alfa')

In [39]:
df_make_origin

0                                    make         country
0                                 AC Cars  United Kingdom
1                                   Acura           Japan
2            Aixam-Mega (including Arola)          France
3                                    Alfa           Italy
4                                  Alpine          France
..                                    ...             ...
234              Austin (included in BMC)  United Kingdom
235               Riley (included in BMC)  United Kingdom
236            Wolseley (included in BMC)  United Kingdom
237            Standard (included in BMC)  United Kingdom
238  Talbot (included in Chrysler Europe)  United Kingdom

[237 rows x 2 columns]

In [63]:
df_autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46405 entries, 0 to 46404
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   mileage    46405 non-null  int64  
 1   make       46405 non-null  object 
 2   model      46262 non-null  object 
 3   fuel       46405 non-null  object 
 4   gear       46223 non-null  object 
 5   offerType  46405 non-null  object 
 6   price      46405 non-null  int64  
 7   hp         46376 non-null  float64
 8   year       46405 non-null  int64  
dtypes: float64(1), int64(3), object(5)
memory usage: 3.2+ MB


In [40]:
df_join_country= pd.merge(df_autos,df_make_origin,how='left',on='make')

In [62]:
df_join_country.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46405 entries, 0 to 46404
Data columns (total 10 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   mileage    46405 non-null  int64 
 1   make       46405 non-null  object
 2   model      46297 non-null  object
 3   fuel       46405 non-null  object
 4   gear       46236 non-null  object
 5   offerType  46405 non-null  object
 6   price      46405 non-null  int64 
 7   hp         46380 non-null  object
 8   year       46405 non-null  int64 
 9   country    46405 non-null  object
dtypes: int64(3), object(7)
memory usage: 3.5+ MB


as an interesting note, the hp vairalbe was converted from a float to object. also the count increased from 46376 to 46380. I really am not all that interested in this but it's messing with my correlation analysis because I can't run an object data type in an array calculation.

In [41]:
df_join_country.head()

mileage        make   model      fuel    gear offerType  price     hp   
0   235000         BMW     316    Diesel  Manual      Used   6800  116.0  \
1    92800  Volkswagen    Golf  Gasoline  Manual      Used   6877  122.0   
2   149300        SEAT    Exeo  Gasoline  Manual      Used   6900  160.0   
3    96200     Renault  Megane  Gasoline  Manual      Used   6950  110.0   
4   156000     Peugeot     308  Gasoline  Manual      Used   6950  156.0   

   year  country  
0  2011  Germany  
1  2011  Germany  
2  2011    Spain  
3  2011   France  
4  2011   France

In [42]:
df_join_country.country.isnull().value_counts()

country
False    46269
True       136
Name: count, dtype: int64

In [43]:
df_join_country.shape

(46405, 10)

In [44]:
df_join_country[df_join_country.make=='Alfa']

mileage  make            model      fuel       gear       offerType   
1458    120000  Alfa  Romeo Giulietta  Gasoline     Manual            Used  \
2331    181382  Alfa  Romeo Giulietta  Gasoline     Manual            Used   
2981     80000  Alfa  Romeo Giulietta  Gasoline     Manual            Used   
3774        10  Alfa  Romeo Giulietta  Gasoline     Manual  Pre-registered   
4464    143000  Alfa       Romeo MiTo  Gasoline     Manual            Used   
...        ...   ...              ...       ...        ...             ...   
45931       50  Alfa    Romeo Stelvio  Gasoline  Automatic  Pre-registered   
45932       50  Alfa    Romeo Stelvio  Gasoline  Automatic  Pre-registered   
46181      500  Alfa    Romeo Stelvio  Gasoline  Automatic   Demonstration   
46212      500  Alfa    Romeo Stelvio  Gasoline  Automatic   Demonstration   
46286       10  Alfa     Romeo Giulia  Gasoline  Automatic            Used   

       price     hp  year country  
1458    6000  120.0  2011   Italy  
2331    5499  120.0  2013   Italy  
2981    8750  105.0  2014   Italy  
3774   18980  120.0  2021   Italy  
4464    5400  135.0  2012   Italy  
...      ...    ...   ...     ...  
45931  45990  280.0  2021   Italy  
45932  46500  280.0  2021   Italy  
46181  44690  200.0  2021   Italy  
46212  55290  280.0  2021   Italy  
46286  88400  510.0  2021   Italy  

[132 rows x 10 columns]

In [45]:
df_join_country[df_join_country.country.isnull()].head(50)


mileage            make             model               fuel   
365       3500          Others               NaN           Electric  \
1985     21800        Microcar               Due             Diesel   
2001      2900          Others               NaN           Electric   
2843        25            Baic               NaN           Gasoline   
3507     39500             9ff               NaN           Gasoline   
5147     21500          Abarth               500           Gasoline   
5308     73837              DS  Automobiles DS 4           Gasoline   
5690     15000          Abarth  595 Competizione           Gasoline   
7359         5          Abarth               500           Gasoline   
7360         5          Abarth               500           Gasoline   
7512     89900  Caravans-Wohnm               NaN             Diesel   
7597         5           Aixam              City             Diesel   
7900     96284              DS  Automobiles DS 3             Diesel   
8518     61600          Abarth               500           Gasoline   
9264         5           Aixam               NaN           Electric   
9432        50          Others               NaN           Gasoline   
10171    84510          Abarth               595           Gasoline   
10366    64219          Abarth       595 Turismo           Gasoline   
10900     3000          Zhidou               D2S           Electric   
11753       90         Maybach           Pullman           Gasoline   
12019     1820          Others               NaN             Diesel   
12236    49670        Microcar              M.Go             Diesel   
13030      800          Others               NaN             Others   
13841    13680        Microcar              M.Go             Diesel   
14602   168000          Abarth         Punto EVO           Gasoline   
14675   312000      Trucks-Lkw              Ford             Diesel   
15323    33000          Abarth               500           Gasoline   
15500   103096              DS  Automobiles DS 5             Diesel   
15765    46100             RAM              1500                LPG   
16429     3400         Maybach           Pullman           Gasoline   
17752      150          Others               NaN           Electric   
18478    27223          Alpina                B3           Gasoline   
18642        1          Others               NaN           Gasoline   
19402    66900         Maybach               NaN           Gasoline   
19742    74800              DS  Automobiles DS 3           Gasoline   
19920    52000          Abarth               500           Gasoline   
20296   100524      Brilliance               BS4                LPG   
21032    67541          Abarth               595           Gasoline   
21544       20          Abarth               595           Gasoline   
21547       20          Abarth               595           Gasoline   
21553       20          Abarth       595 Turismo           Gasoline   
21555       20          Abarth              595C           Gasoline   
22164    30180          Abarth               695           Gasoline   
22281    53500        Microcar               Due             Diesel   
22457     7500          FISKER             Karma  Electric/Gasoline   
22792       10          Abarth               595           Gasoline   
23220       20          Abarth               595           Gasoline   
23815    53100          Alpina                D3             Diesel   
24203   108500          Alpina                D3             Diesel   
24277       10          Abarth               595           Gasoline   

            gear       offerType   price     hp  year country  
365       Manual            Used    5500   38.0  2018     NaN  
1985   Automatic            Used    8490    5.0  2017     NaN  
2001   Automatic            Used    7900    5.0  2018     NaN  
2843   Automatic            Used   20900  136.0  2018     NaN  
3507      Manual            Used    7000   20.0  2018     NaN  
5

In [46]:
df_join_country[df_join_country.make=="Abarth"]=df_join_country[df_join_country.make=="Abarth"].fillna('Italy')

In [47]:
df_join_country[df_join_country.country.isnull()]= df_join_country[df_join_country.country.isnull()].fillna('undetermined')

In [48]:
df_join_country[df_join_country.country=='undetermined'].country.value_counts()

country
undetermined    93
Name: count, dtype: int64

In [49]:
df_join_country[df_join_country.country == 'undetermined'].tail(10)

mileage              make                       model   
43610      700            Others                undetermined  \
43612      300            Others                undetermined   
44072        1  Trailer-Anhänger                undetermined   
45559      100                DS  Automobiles DS 7 Crossback   
45571    14000                DS  Automobiles DS 7 Crossback   
46074        1            Others                undetermined   
46102        1            Others                undetermined   
46234     1800                DS  Automobiles DS 7 Crossback   
46334      100              DFSK                   Glory 580   
46352     1000            Alpina                          B3   

                    fuel          gear       offerType  price     hp  year   
43610           Electric  undetermined            Used   6300    5.0  2019  \
43612           Electric     Automatic            Used   7000    5.0  2019   
44072             Others  undetermined            Used   2500    1.0  2019   
45559             Diesel     Automatic            Used  53995  177.0  2021   
45571  Electric/Gasoline     Automatic            Used  59995  299.0  2021   
46074           Gasoline        Manual  Pre-registered  23990  150.0  2021   
46102           Gasoline        Manual  Pre-registered  23990  150.0  2021   
46234  Electric/Gasoline     Automatic            Used  53690  299.0  2021   
46334           Gasoline     Automatic            Used  19990  145.0  2021   
46352           Gasoline     Automatic   Demonstration  89900  462.0  2021   

            country  
43610  undetermined  
43612  undetermined  
44072  undetermined  
45559  undetermined  
45571  undetermined  
46074  undetermined  
46102  undetermined  
46234  undetermined  
46334  undetermined  
46352  undetermined

In [50]:
df_join_country.country.value_counts()

country
Germany           20875
France             5029
United States      4882
Japan              3825
South Korea        2980
Czech              2889
Spain              1989
Italy              1935
Sweden              808
Romania             715
United Kingdom      351
undetermined         93
Russia               33
Netherlands           1
Name: count, dtype: int64

In [51]:
df_join_country.shape

(46405, 10)

In [52]:
df_join_country[df_join_country.country=='Germany'].price.mean()

19234.046562874253

In [53]:
df_join_country[df_join_country.country == 'Spain'].price.mean()

13777.675716440423

In [54]:
df_join_country[df_join_country.country == 'United Kingdom'].price.mean()

69797.99715099715

In [55]:
df_join_country.price.mean()

16572.33722659196

In [56]:
df_join_country.shape

(46405, 10)

In [58]:
df_join_country.describe()

mileage         price          year
count  4.640500e+04  4.640500e+04  46405.000000
mean   7.117786e+04  1.657234e+04   2016.012951
std    6.262531e+04  1.930470e+04      3.155214
min    0.000000e+00  1.100000e+03   2011.000000
25%    1.980000e+04  7.490000e+03   2013.000000
50%    6.000000e+04  1.099900e+04   2016.000000
75%    1.050000e+05  1.949000e+04   2019.000000
max    1.111111e+06  1.199900e+06   2021.000000

In [57]:
#df_join_country.to_pickle('/Users/pkc/PycharmProjects/pythonProject/Automotive/df_join_lists.pickle')